src/plan_expand_hypertable.c

```diff
@@ -139,6 +193,63 @@ ts_plan_expand_hypertable_valid_hypertable(Hypertable *ht, Query *parse, Index r
 	return true;
 }
 
+/**
+ * Get chunk oids from either restrict info or explicit chunk exclusion. Explicit chunk exclusion
+ * takes precedence.
+ */
+static List *
+get_chunk_oids(CollectQualCtx *ctx, PlannerInfo *root, RelOptInfo *rel, Hypertable *ht)
+{
+	List	   *chunk_oids = NIL;
+
+	if (ctx->chunk_exclusion_func == NULL)
+	{
+		HypertableRestrictInfo *hri = ts_hypertable_restrict_info_create(rel, ht);
+
+		/*
+		 * This is where the magic happens: use our HypertableRestrictInfo
+		 * infrastructure to deduce the appropriate chunks using our range
+		 * exclusion
+		 */
+		ts_hypertable_restrict_info_add(hri, root, ctx->restrictions);
+		chunk_oids = find_children_oids(hri, ht, AccessShareLock);
+	}
+	else
+	{
+		/* explicit chunk exclusion */
+		Const	   *chunks_arg;
+		ArrayIterator chunk_id_iterator;
+		Datum		elem = (Datum) NULL;
+		bool		isnull;
+
+		Assert(ctx->chunk_exclusion_func->args->length == 2);
+		chunks_arg = (Const *) lsecond(ctx->chunk_exclusion_func->args);
+
+		if (chunks_arg->constisnull)
+			elog(ERROR, "chunks to filter on can't be NULL");
+
+		chunk_id_iterator = array_create_iterator(DatumGetArrayTypeP(chunks_arg->constvalue), 0, NULL);
+
+
+		while (array_iterate(chunk_id_iterator, &elem, &isnull))
+		{
+			if (!isnull)
+			{
+				int32		chunk_id = DatumGetInt32(elem);
+				Chunk	   *chunk = ts_chunk_get_by_id(chunk_id, 0, true);
+
+				if (chunk->fd.hypertable_id != ht->fd.id)
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("chunk id %d does not belong to the hypertable %s", chunk_id, NameStr(ht->fd.table_name))));
+
+				chunk_oids = lappend_int(chunk_oids, chunk->table_id);
+			}
+		}
```

Comment:
Think you need and `array_free_iterator` here.

Reply:
Since it's allocated in `CurrentMemoryContext` isn't it going to be freed automatically by PostgreSQL?

Reply:
Yes, but given that there is an explicit array API for this I think it is good to follow the pattern.